00XX-common-ledger-protocol/00XX-common-ledger-protocol.md

```diff
@@ -0,0 +1,273 @@
+  sent.
+
+- **CLP Packets** are the messages defined in the CLP ASN.1 spec.
+
+- **Peers** are the parties on a CLP connection. Your peer is the party on the
+  other side of the CLP connection.
+
+- The **Bilateral Ledger** is the ledger which a CLP connection controls.  The
+  bilateral ledger is a persistent log of CLP packets, which can be used to
+deduce the current balance between two peers and the state of all transfers
+between them.
+
+- **Authoritative State** means that the entire Bilateral Ledger is being
+  tracked by a party. Both peers on a CLP connection can keep authoritative
+state, which means they can get into dispute if they disagree on the state of a
+transfer. This usually happens when network latency causes the peers to
+disagree about expiries. If one party keeps authoritative state, the other
+party must trust them not to tamper with it.
+
+## Overview
+
+CLP is broken up into 4 different RPC requests, which can get 2 different
+responses. Every CLP packet follows a common structure:
+
+```
+  +---------------+
+1 | Type (1)      |
+  +---------------+
+1 | Request ID    |
+2 | (2)           |
+3 |               |
+4 |               |
+  +---------------+
+1 | Length Prefix |
+2 | (3)           |
+  +---------------+
+  | Packet-       |
+  | Specific      |
+  | Data          |
+  . (4)           |
+  .               |
+  .               |
+  |               |
+  +---------------+
+1 | Side Protocol |
+2 | Count (5)     |
+  +---------------+
+  | Side Protocol |
+  | Data          |
+  . (6)           |
+  .               |
+  .               |
+  |               |
+  +---------------+
+```
+
+1. **Type**: A 1-byte value describing what type of CLP packet this is.
+The [CLP ASN.1 spec] describes what the different type IDs mean.
+
+2. **Request ID**: A 4-byte value used to correlate this requests and
+responses.
+
+3. **Length Prefix**: A 1-byte (if under 128) or 2-byte value, containing the
+combined length of the packet-specific data and side protocol data sections.
+
+4. **Packet-Specific Data:** Fields specific to the type of CLP packet. Variable
+length.
+
+5. **Side Protocol Count:** Variable-length integer containing the number of
+side protocols carried by this packet.
+
+5. **Side Protocol Data:** A list of key-value pairs, where the key is a
+length-prefixed ascii string, and the value is a length-prefixed octet string.
+The key signifies the side protocol's name, and the value is the data.
+
+## Flow
+
+CLP uses a simple RPC flow. A request-type CLP packet is sent, and a
+response-type CLP packet is sent in response with the same request ID.
+
+If a malformed CLP packet is sent, there should be no response. If a
+response-type CLP packet is sent without matching a previous request, it should
+be ignored.  If a request has already been responded to and another response is
+sent, the second response should be ignored.
+
+Because it would be too slow to atomically save all requestIds that are
```

Comment:
This could use some clarification. Transfers must be idempotent, but it doesn't matter if other requests aren't

Reply:
Transfers are idempotent with respect to the transferId, but that's different from RPC-level idempotency

Reply:
I understood that after reading through the whole doc but while reading this section I thought it was saying that nothing was idempotent

Reply:
This ambiguity comes from the fact that requestId and transferId are at different layers in this stack.

`requestId` is at the very bottom layer of the CLP stack and is simply for matching request/responses. There is no application logic that looks at these, in fact they could be invisible to the code that processes the actual message as long as it gets the correct response for its request.

`transferId` is related to business logic (a layer up in the CLP stack) where idempotency is enforced.

I think it would be less confusing if messages also had an ID that was relevant to the business logic layer.